### PR TITLE
chg: test: Use travis-ci for testing.

### PR DESCRIPTION
**Pull request by ashmastaflash**
Set up travis-ci for integration testing with Halo.

Re-factor in halo.py for better performance, reliability, consistency.

Re-factor in containerized.py for reduced duplication, created a generic
container launch method for simplifying the process of adding newer
functionality.

Changed behavior of get_id_for_server_target() and
get_id_for_group_target() to return a consistent format.
Re-worked dependent functions to accommodate new
functionality.

### CODE REVIEW

File: apputils/containerized.py [hunk 1]
```diff
@@ -18,107 +20,117 @@ def __init__(self):
...
+                                                           image))
+        result = self.client.containers.run(image, name=container_name,
+                                            detach=False,
+                                            mem_limit=self.mem_limit,
+                                            environment=env_vars)
+        self.client.containers.get(container_name).remove()
+        return result
+
+    @classmethod
+    def generate_random_name(cls):
```
Comment (eatwithforks) [inline apputils/containerized.py:62]: Consider using python module 'uuid' to generate a true unique name.
Make a UUID based on the host ID and current time:

import uuid

id = str(uuid.uuid1())
print id

output: '14b83238-5534-11e8-9b1d-f45c8990a101'

File: apputils/containerized.py [hunk 2]
```diff
@@ -18,107 +20,117 @@ def __init__(self):
...
-            self.client.containers.get(container_name).remove()
-        except docker.errors.APIError:
-            pass
-        result = self.client.containers.run(image, name=container_name,
-                                            detach=False,
-                                            mem_limit=self.mem_limit,
-                                            environment=environment)
-        self.client.containers.get(container_name).remove()
+        env_vars = {"TARGET_DATE": target_date,
+                    "AWS_S3AWS_S3_BUCKET": s3_bucket_name}
```
Comment (eatwithforks) [inline apputils/containerized.py]: Typo: "AWS_S3AWS_S3_BUCKET" should be "AWS_S3_BUCKET"

File: apputils/containerized.py [hunk 3]
```diff
@@ -18,107 +20,117 @@ def __init__(self):
...
-            "HALO_API_KEY": self.halo_key,
-            "HALO_API_SECRET_KEY": self.halo_secret,
-            "TARGET_DATE": target_date,
-            "AWS_S3_BUCKET": s3_bucket_name,
-            "AWS_ACCESS_KEY_ID": self.aws_key,
-            "AWS_SECRET_ACCESS_KEY": self.aws_secret,
-            "HTTPS_PROXY": self.https_proxy
-        }
+        env_vars = {"TARGET_DATE": target_date,
+                    "AWS_S3AWS_S3_BUCKET": s3_bucket_name}
```
Comment (eatwithforks) [inline apputils/containerized.py]: Typo

File: apputils/halo.py [hunk 4]
```diff
@@ -16,52 +16,67 @@ def __init__(self):
...
-                result = fmt.format_item(facts, "server_ec2")
-            else:
-                result = fmt.format_item(facts, "server_facts")
-            print("ServerReport: Getting server issues")
-            result += fmt.format_list(self.get_issues_by_server(server_id),
-                                      "issue")
-            print("ServerReport: Getting server events")
-            result += fmt.format_list(self.get_events_by_server(server_id),
-                                      "event")
+        if len(server_ids) == 0:
```
Comment (eatwithforks) [inline apputils/halo.py]: Prefer guard clause over nested conditionals.

Example:

server_ids = self.get_id_for_server_target(target)
if not server_ids:
&nbsp;&nbsp;&nbsp;&nbsp;return "Unable to find server %s" % target

perform below loop without the else.

File: apputils/containerized.py (hunk 1 above)
Comment (ashmastaflash) [reply apputils/containerized.py:62]: Great idea, I'll do this instead.

File: apputils/containerized.py (hunk 2 above)
Comment (ashmastaflash) [reply apputils/containerized.py]: JyeCorrect > AutoCorrect  Fix coming right up.

File: apputils/halo.py (hunk 4 above)
Comment (ashmastaflash) [reply apputils/halo.py]: OK, NP.

File: apputils/containerized.py (hunk 3 above)
Comment (ashmastaflash) [reply apputils/containerized.py]: same as above

File: run-tests.sh [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+if [ "$RUN_INTEGRATION_TESTS" = "True" ] ; \
+    then echo "Run all tests" && \
+        python -mpy.test \
+            --cov-report term-missing \
+            --cov=apputils \
+            /app/halocelery/tests/style \
+            /app/halocelery/tests/unit \
+            /app/halocelery/tests/functional ;  \
+    else echo Not running integration tests!. && \
+        python -mpy.test \
```
Comment (eatwithforks) [inline run-tests.sh]: typo: python -mpy needs a space in between.

File: run-tests.sh [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+    pytest-cover==3.0.0 \
+    pytest-flake8==0.6 \
+    pyflakes==1.2.3 \
+    flake8==3.0.0
+
+# If RUN_INTEGRATION_TESTS is set, run integration tests.
+echo "Run integration tests? ${RUN_INTEGRATION_TESTS}"
+if [ "$RUN_INTEGRATION_TESTS" = "True" ] ; \
+    then echo "Run all tests" && \
+        python -mpy.test \
```
Comment (eatwithforks) [inline run-tests.sh]: typo

Comment (ashmastaflash) [reply run-tests.sh]: For what it's worth, it's not uncommon to use Python's `-m` argument without a space between the `-m` and the module being called.  See the example from Matplotlib: https://matplotlib.org/faq/installing_faq.html (look for `python -mpip install matplotlib`)

But it works either way.

File: apputils/containerized.py [hunk 7]
```diff
@@ -18,107 +19,116 @@ def __init__(self):
...
-        }
-
+        env_expand = {"HALO_API_KEY": "HALO_API_KEY",
+                      "HALO_API_SECRET_KEY": "HALO_API_SECRET_KEY",
+                      "AWS_ACCESS_KEY_ID": "AWS_ACCESS_KEY_ID",
+                      "AWS_SECRET_ACCESS_KEY": "AWS_SECRET_ACCESS_KEY",
+                      "HTTPS_PROXY": "HTTPS_PROXY"}
+        env_vars = {"OUTPUT_FORMAT": "csv"}
         # Populate optional fields to support multi-account inventory.
         optional_fields = ["AWS_ROLE_NAME", "AWS_ACCOUNT_NUMBERS"]
```
Comment (eatwithforks) [inline apputils/containerized.py:89]: consider doing this without the loop logic. can you add optional fields to env vars and have it handled in expand_and_update function? Having a loop logic in here handling fields doesn't follow single responsibility principle.

File: apputils/containerized.py [hunk 8]
```diff
@@ -18,107 +19,116 @@ def __init__(self):
...
-        environment = {
-            "HALO_API_KEY": self.halo_key,
-            "HALO_API_SECRET_KEY": self.halo_secret,
-            "AWS_ACCESS_KEY_ID": self.aws_key,
-            "AWS_SECRET_ACCESS_KEY": self.aws_secret,
-            "HTTPS_PROXY": self.https_proxy,
-            "OUTPUT_FORMAT": "csv"
-        }
-
+        env_expand = {"HALO_API_KEY": "HALO_API_KEY",
```
Comment (eatwithforks) [inline apputils/containerized.py:82]: Applies to all: consider having a yaml file that includes the repo names and the env variables they require. That hash can be read by a singular function to set all these images up. 

Right now, this method for declaring hashes in each function is failing the open-closed principle. It is impossible to extend functionality without modifying the current code.

**Comment by eatwithforks** [comment]
+1. do  interface segregation for env variables in another ticket

**Pull request merged**